director/projects/api/views/nodes.py

```diff
@@ -0,0 +1,222 @@
+        - For `text/html` (and others) returns a very simple HTML rendering of the
+        of the node (only type etc) if the user does not have read access to the project.
+        Otherwise, should return a full HTML rendering of the node using Encoda.
+        """
+        node = get_object_or_404(Node, key=key)
+        accept = request.META.get("HTTP_ACCEPT")
+        if format == "json" or (accept and "application/json" in accept):
+            # Require the user is authenticated and has VIEW permissions for the project
+            if not request.user.is_authenticated:
+                raise NotAuthenticated
+            if not self.is_permitted(
+                request.user, ProjectPermissionType.VIEW, pk=node.project.id
+            ):
+                raise PermissionDenied
+
+            serializer = NodeSerializer(node, context={"request": request})
+            return Response(serializer.data)
+        else:
+            if self.is_permitted(
+                request.user, ProjectPermissionType.VIEW, pk=node.project.id
+            ):
+                # Return a more complete view if the user has VIEW permissions.
+                # This should include public projects.
+                try:
+                    # Currently allow this to fail if the converter binary
+                    # can not be found e.g. during CI testing
+                    conversion = ConverterFacade().convert(
+                        input_data=ConverterIo(
+                            ConverterIoType.PIPE,
+                            json.dumps(node.json).encode("utf8"),
+                            ConversionFormatId.json,
+                        ),
+                        output_data=ConverterIo(
+                            ConverterIoType.PIPE, None, ConversionFormatId.html
+                        ),
+                        context=ConverterContext(standalone=False),
+                    )
+                    html = conversion.stdout.decode("utf8")
+                except FileNotFoundError:
+                    html = ""
+
+                app_url, app_name = None, None
+                if node.app == "stencila:encoda":
+                    app_name = "Stencila Encoda"
+                    app_url = "https://github.com/stencila/encoda#readme"
+                elif node.app == "stencila:gsuita":
+                    app_name = "Stencila for GSuite"
+                    app_url = "https://gsuite.google.com/marketplace/app/stencila/110435422451"
+
+                return Response(
+                    {
+                        "node_type": node_type(node.json),
+                        "app_url": app_url,
+                        "app_name": app_name,
+                        "node": node,
+                        "html": html,
+                    },
+                    template_name="projects/node_complete.html",
+                )
+            else:
+                # Return a basic view if the user does NOT have VIEW permissions.
```

Comment:
Should just raise permission denied?

Reply:
No because we want Google's scraper (and possibly others) to be able to fetch the HTML and generate the preview (see the related issue in GSuita) that indicates this is a Stencila CodeChunk (or whatever). The node `key` is intended to be kept a secret and the project permissions checking is just an additional layer in case it is leaked (in which case someone will only see that what type of node it is, nothing else).